### PR TITLE
Add simple blockchain commitment API

### PR DESCRIPTION
**Pull request by MsRandom**
New base types:
- DataFieldType: defines what type a commitment field is using(whether Raw data or a type of hash)
- CommitmentDataField: A field to be committed or queried from the chain
- CommitmentQuery: The full result of the commitment query, which includes all committed fields, the block they were committed at, and the deposit made by the hotkey to commit them

New base functions:
- set_commitment: Commit a number of fields to the chain for a set netuid and wallet(fields are list[CommitmentDataField])
- query_commitment: Queries commitment made with set_commitment, returns CommitmentQuery

Helper types:
- RawCommitment: Includes only the data of one field with raw data and not a hash

Helper functions:
- publish_raw_commitment: Commit a single byte-string to the chain as raw data
- get_raw_commitment: Query existing single byte-string commitment

### CODE REVIEW

File: fiber/chain_interactions/commitments.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    BLAKE_TWO_256 = "BlakeTwo256"
+    SHA_256 = "Sha256"
+    KECCAK_256 = "Keccak256"
+    SHA_THREE_256 = "ShaThree256"
+
+
+CommitmentDataField: TypeAlias = tuple[DataFieldType, bytes] | None
+
+
+class CommitmentQuery(BaseModel):
```
Comment (namoray) [inline fiber/chain_interactions/commitments.py]: in /chain_interactions (soon to be renamed /chain) there is a models.py where these could live

File: fiber/chain_interactions/commitments.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    deposit: int
+
+
+class RawCommitment(BaseModel):
+    data: bytes
+    block: int
+    deposit: int
+
+
+def _serialize_field(field: CommitmentDataField):
```
Comment (namoray) [inline fiber/chain_interactions/commitments.py]: return typehint missing

File: fiber/chain_interactions/commitments.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    if not field:
+        return {str(None): b''}
```
Comment (namoray) [inline fiber/chain_interactions/commitments.py]: str(None) 0.o 

should probably be a constant in fiber.constants.py or similar, of 'None' if we want to return that [im guessing this is straight from BT?]

File: fiber/chain_interactions/commitments.py (hunk 2 above)
Comment (namoray) [inline fiber/chain_interactions/commitments.py]: I'd also call field (commitment_data_field) but i like my verbose names :D

File: fiber/chain_interactions/commitments.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+
+def _deserialize_field(field: dict[str, bytes | str]) -> CommitmentDataField:
+    data_type, data = field.items().__iter__().__next__()
+
+    if data_type == str(None):
+        return None
+
+    if data_type.startswith(DataFieldType.RAW.value):
+        expected_length = int(data_type[len(DataFieldType.RAW.value):])
```
Comment (namoray) [inline fiber/chain_interactions/commitments.py]: expected_data_field_length... is that too verbose?

File: fiber/chain_interactions/commitments.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+        data_type = DataFieldType.RAW.value
+        data = bytes.fromhex(data[2:])
+
+        if len(data) != expected_length:
+            raise ValueError(f"Got commitment raw field expecting {expected_length} data but got {len(data)} data")
+
+    return DataFieldType(data_type), data
```
Comment (namoray) [inline fiber/chain_interactions/commitments.py]: Could we explicitly cast it to the right type

File: fiber/chain_interactions/commitments.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+    wait=wait_exponential(multiplier=1, min=1, max=4),
+    reraise=True,
+)
+def _query_commitment(
+    substrate: SubstrateInterface,
+    netuid: int,
+    hotkey: str,
+    block: int | None = None,
+) -> ScaleType:
+    return substrate.query(
```
Comment (namoray) [inline fiber/chain_interactions/commitments.py]: substrate is better than substrate_interface though, i will replace it to that everywhere thanks.

Note for this, you may need to close the connection after or reinstantiate if it fails. I'm going to try to make a fixed SubstrateInterface class for this, which handles the constant SSL errors we're seeing

Comment (namoray) [reply fiber/chain_interactions/commitments.py]: so basically that was a note to self ^

File: fiber/chain_interactions/commitments.py [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+
+    if not wait_for_finalization and not wait_for_inclusion:
+        return True, "Not waiting for finalization or inclusion."
+
+    response.process_events()
+
+    if response.is_success:
+        return True, "Successfully submitted commitment."
+
+    return False, format_error_message(response.error_message)
```
Comment (namoray) [inline fiber/chain_interactions/commitments.py]: What's ever done with this error messaeg? can we just logger.error() it instead, and then just return a bool? makes life easier

File: fiber/chain_interactions/weights.py [hunk 8]
```diff
@@ -140,7 +128,7 @@ def _send_extrinsic(
         if response.is_success:
             return True, "Successfully set weights."
 
-        return False, _format_error_message(response.error_message)
+        return False, format_error_message(response.error_message)
```
Comment (namoray) [inline fiber/chain_interactions/weights.py]: NTS: i need to get rid of this log the error and move on too

**Review by namoray — APPROVED** [review]

File: fiber/chain_interactions/commitments.py (hunk 3 above)
Comment (MsRandom) [reply fiber/chain_interactions/commitments.py]: Yep this is the field data type `None`, moved it to constants

**Review by namoray — APPROVED** [review]

**Pull request merged**